### DISCUSSION TITLE
Coloured polyline and cached tiles

### DISCUSSION THREAD

**Discussion by Haukinger** [Mapsui/Mapsui#2815] [General]
I'm looking to replace Osmdroid in my net android app, and the main features I need are polylines (LineString?) that are coloured per point, and caching of tiles, so that each tile is downloaded only once (given the cache is large enough). Can I do this with Mapsui?

**Comment by pauldendulk** [comment]
Caching of tiles is possible.

Styling a LineString so that every point has a different color is not possible out of the box. Options:
- Use a custom style renderer. https://mapsui.com/v5/custom-style-renders/
- Extract the Points from the LineString and put them in a separate layer with their own layer using a ThemeStyle.

I wonder how that works in Osmdroid, how do you define a style where each point has a different color if you do not know how many points are in the LineString? Is it a gradient?

If you just need the Points and LineString to have a different colors then this is possible out of the box by setting a SymboStyle on a VectorStyle (I think this is only in V5).

**Reply by Haukinger** [reply]
With osmdroid, I can add "metadata" (a float in my case) to each point and set up a color mapping from that to a range of colors, i.e. I create a `Polyline` and add `PolychromaticPaintList` - in fact, that looks quite similar to a custom style, just built in.

**Reply by pauldendulk** [reply]
Adding metadata to individual points of a LineString seems an unusual thing to do. I have not worked with software that works like that. If individual points have some kind of identity they I would think they should be regular Points. If you want to draw a line to connect them then draw a LineString underneath. What kind of data is this? What does it represent in the real world?

**Reply by Haukinger** [reply]
In my osmdroid app, I color the points by speed - each point is assigned the percentage of maximum speed on the track and my color mapping makes slow parts appear red, medium parts yellow and the fastest parts green. Right now I'm experimenting with mapsui on wpf for a similar problem, and I use `GeometryFeature.Data` to add meta data (curvature) for each point with a custom style to color the dots.

**Reply by charlenni** [reply]
I think, that this is a common case. In OpenGL each point has a color. With that a triangle is drawn with a gradient between the points.

It is used often in maps apps to make, as mentioned before, the speed, the heigth or the gradient of a track more visible.

The easiest way to do this would be to derive from LineString, save the color in the data of the points (extra list?). Then create an own renderer for this type and use a shader while drawing the line.
